src/pretix/base/models/orders.py

```diff
@@ -1841,9 +1845,14 @@ def _mark_order_paid(self, count_waitinglist=True, send_mail=True, force=False,
             cancellations = self.order.invoices.filter(is_cancellation=True).count()
             gen_invoice = (
                 (invoices == 0 and self.order.event.settings.get('invoice_generate') in ('True', 'paid')) or
-                0 < invoices <= cancellations
+                0 < invoices <= cancellations or
+                self.order.invoice_dirty
             )
             if gen_invoice:
+                if invoices:
+                    last_i = self.order.invoices.filter(is_cancellation=False).last()
+                    if not last_i.refered.exists():
```

Comment:
See comment below regarding refered being exclusively cancelations. Might be that case in reality, but code comments IMHO do not read that way. So maybe more robust

```suggestion
                    if not last_i.canceled:
```